src/ray/core_worker/core_worker.cc

```diff
@@ -1783,31 +1794,51 @@ Status CoreWorker::GetLocationFromOwner(
     return Status::OK();
   }
 
+  absl::flat_hash_map<rpc::Address, std::vector<ObjectID>> objects_by_owner;
+  for (const auto &object_id : object_ids) {
+    rpc::Address owner_address;
+    RAY_RETURN_NOT_OK(GetOwnerAddress(object_id, &owner_address));
+    objects_by_owner[owner_address].push_back(object_id);
+  }
+
   auto mutex = std::make_shared<absl::Mutex>();
-  auto num_remaining = std::make_shared<size_t>(object_ids.size());
+  auto num_remaining = std::make_shared<size_t>(objects_by_owner.size());
   auto ready_promise = std::make_shared<std::promise<void>>();
   auto location_by_id =
       std::make_shared<absl::flat_hash_map<ObjectID, std::shared_ptr<ObjectLocation>>>();
 
-  for (const auto &object_id : object_ids) {
-    rpc::Address owner_address;
-    RAY_RETURN_NOT_OK(GetOwnerAddress(object_id, &owner_address));
+  for (const auto &pair : objects_by_owner) {
+    const auto &owner_address = pair.first;
+    const auto &owner_object_ids = pair.second;
```

Comment:
I think you can do

```
for (const auto &[owner_address, owner_object_ids]: objects_by_owner)
```

Reply:
This is my original implementation but failed at CI, it seems that Structured bindings (introduced in C++17) cannot be directly captured by a lambda until C++20, and even then. So I move it back to define them explicitly.
error from ci 
```
[2024-03-26T12:45:33Z] #13 160.1     src/ray/core_worker/core_worker.cc:1821:6: error: 'owner_object_ids' in capture list does not name a variable
[2024-03-26T12:45:33Z] #13 160.1         [owner_object_ids, mutex, num_remaining, ready_promise, location_by_id, this, owner_address](
[2024-03-26T12:45:33Z] #13 160.1          ^
[2024-03-26T12:45:33Z] #13 160.1     src/ray/core_worker/core_worker.cc:1821:83: error: 'owner_address' in capture list does not name a variable
[2024-03-26T12:45:33Z] #13 160.1         [owner_object_ids, mutex, num_remaining, ready_promise, location_by_id, this, owner_address](
[2024-03-26T12:45:33Z] #13 160.1                                                                                       ^
[2024-03-26T12:45:33Z] #13 160.1     src/ray/core_worker/core_worker.cc:1826:34: error: reference to local binding 'owner_object_ids' declared in enclosing function 'ray::core::CoreWorker::GetLocationFromOwner'
[2024-03-26T12:45:33Z] #13 160.1               for (size_t i = 0; i < owner_object_ids.size(); ++i) {
[2024-03-26T12:45:33Z] #13 160.1                                      ^
[2024-03-26T12:45:33Z] #13 160.1     src/ray/core_worker/core_worker.cc:1810:36: note: 'owner_object_ids' declared here
[2024-03-26T12:45:33Z] #13 160.1       for (const auto& [owner_address, owner_object_ids] : objects_by_owner) {
[2024-03-26T12:45:33Z] #13 160.1                                        ^
[2024-03-26T12:45:33Z] #13 160.1     src/ray/core_worker/core_worker.cc:1829:19: error: reference to local binding 'owner_object_ids' declared in enclosing function 'ray::core::CoreWorker::GetLocationFromOwner'
[2024-03-26T12:45:33Z] #13 160.1                       owner_object_ids[i], std::make_shared<ObjectLocation>(locations[i]));
[2024-03-26T12:45:33Z] #13 160.1                       ^
[2024-03-26T12:45:33Z] #13 160.1     src/ray/core_worker/core_worker.cc:1810:36: note: 'owner_object_ids' declared here
[2024-03-26T12:45:33Z] #13 160.1       for (const auto& [owner_address, owner_object_ids] : objects_by_owner) {
[2024-03-26T12:45:33Z] #13 160.1                                        ^
[2024-03-26T12:45:33Z] #13 160.1     src/ray/core_worker/core_worker.cc:1833:31: error: reference to local binding 'owner_address' declared in enclosing function 'ray::core::CoreWorker::GetLocationFromOwner'
[2024-03-26T12:45:33Z] #13 160.1                                << owner_address.worker_id()
[2024-03-26T12:45:33Z] #13 160.1                                   ^
[2024-03-26T12:45:33Z] #13 160.1     src/ray/core_worker/core_worker.cc:1810:21: note: 'owner_address' declared here
[2024-03-26T12:45:33Z] #13 160.1       for (const auto& [owner_address, owner_object_ids] : objects_by_owner) {
[2024-03-26T12:45:33Z] #13 160.1                         ^
[2024-03-26T12:45:33Z] #13 160.1     src/ray/core_worker/core_worker.cc:1821:77: error: lambda capture 'this' is not used [-Werror,-Wunused-lambda-capture]
[2024-03-26T12:45:33Z] #13 160.1         [owner_object_ids, mutex, num_remaining, ready_promise, location_by_id, this, owner_address](
[2024-03-26T12:45:33Z] #13 160.1                       
```